File: src/auth/exedev.ts
// The whole of this product's coupling to exe.dev, in one file so it is one
// grep. See ADR 0009 and docs/exedev-identity-header-study.md.
//
// exe.dev terminates TLS at the edge and adds two headers to a request from a
// user it has authenticated:
//
//     X-ExeDev-UserID    a stable, unique user identifier
//     X-ExeDev-Email     the user's email address
//
// They are present only when the user is authenticated. There is no OIDC
// endpoint, no JWT and no userinfo call — headers are the entire interface. The
// documentation writes the names as X-ExeDev-Email in one place and
// X-Exedev-Email in another; node lower-cases every incoming header name, so
// reading them in lower case is right whichever the proxy sends.
//
// WHAT THIS DOES NOT PROVE. The docs say the proxy strips X-Exedev-Authorization
// and that a caller "cannot forge" X-Exedev-Source-Vm. They say nothing of the
// kind about these two. Until the measurement in the study is done, treat the
// email as IDENTIFICATION — it says which account the proxy believes is at the
// keyboard — and not as proof that the request came through the proxy at all.
// The study records why it cannot be measured from inside the VM.

import type { IncomingHttpHeaders } from 'node:http';

/** The path prefix exe.dev reserves. Nothing of ours may live under it. */
export const EXEDEV_PREFIX = '/__exe.dev/';

export const EMAIL_HEADER = 'x-exedev-email';
export const USER_ID_HEADER = 'x-exedev-userid';

export type Identity = {
  email: string;
  /** Absent when the proxy sent an email but no id. Not worth refusing over. */
  userId?: string;
};

/** The identity exe.dev asserts for this request, or null if it asserts none. */
export function identityOf(headers: IncomingHttpHeaders): Identity | null {
  const email = single(headers[EMAIL_HEADER]);
  if (!email) return null;
  return { email, userId: single(headers[USER_ID_HEADER]) };
}

/**
 * Where to send a browser that carries no identity.
 *
 * `redirect` is a path on this same host, and it must stay one: an absolute URL
 * here would be an open redirect through our own login link.
 */
export function loginRedirect(returnTo: string): string {
  const path = returnTo.startsWith('/') && !returnTo.startsWith('//') ? returnTo : '/';
  return `${EXEDEV_PREFIX}login?redirect=${encodeURIComponent(path)}`;
}

/**
 * Compare two email addresses for "is this the household".
 *
 * The domain is case-insensitive and in practice so is every mailbox worth
 * caring about, so both sides are lower-cased. Nothing more clever: guessing at
 * plus-addressing or dots would widen the allow list, and this list has one
 * entry on purpose.
 */
export function sameEmail(one: string, other: string): boolean {
  return one.trim().toLowerCase() === other.trim().toLowerCase();
}

function single(value: string | string[] | undefined): string | undefined {
  const first = Array.isArray(value) ? value[0] : value;
  const trimmed = first?.trim();
  return trimmed ? trimmed : undefined;
}

File: lib/mealplan/auth/exedev.ex
defmodule Mealplan.Auth.Exedev do
  @moduledoc """
  The whole of this product's coupling to exe.dev, in one module so it is one
  grep. Ported from `src/auth/exedev.ts`. See ADR 0009.

  exe.dev terminates TLS at the edge and adds two headers for an authenticated
  user: `X-ExeDev-UserID` (a stable id) and `X-ExeDev-Email`. They are present
  only when the user is authenticated. There is no OIDC endpoint — headers are
  the entire interface.
  """

  @prefix "/__exe.dev/"
  def prefix, do: @prefix

  @email_header "x-exedev-email"
  @user_id_header "x-exedev-userid"

  @type identity :: %{email: String.t(), user_id: String.t() | nil}

  @doc "The identity exe.dev asserts for this `Plug.Conn`, or nil."
  @spec identity_of(Plug.Conn.t()) :: identity() | nil
  def identity_of(conn) do
    case single(Plug.Conn.get_req_header(conn, @email_header)) do
      nil -> nil
      email -> %{email: email, user_id: single(Plug.Conn.get_req_header(conn, @user_id_header))}
    end
  end

  @doc """
  Where to send a browser that carries no identity. `return_to` must be a path
  on this same host — an absolute URL would be an open redirect through our own
  login link.
  """
  def login_redirect(return_to) do
    path =
      if is_binary(return_to) and String.starts_with?(return_to, "/") and
           not String.starts_with?(return_to, "//"),
         do: return_to,
         else: "/"

    "#{@prefix}login?redirect=#{URI.encode_www_form(path)}"
  end

  @doc "Case-insensitive comparison for \"is this the household\"."
  def same_email?(one, other) do
    String.downcase(String.trim(one || "")) == String.downcase(String.trim(other || ""))
  end

  defp single([value | _]) do
    case String.trim(value) do
      "" -> nil
      trimmed -> trimmed
    end
  end

  defp single(_), do: nil
end
